Mathlib/RingTheory/Multiplicity.lean

```diff
@@ -281,6 +281,10 @@ theorem emultiplicity_eq_coe {n : ℕ} :
   · rw [and_imp]
     apply emultiplicity_eq_of_dvd_of_not_dvd
 
+theorem emultiplicity_eq_ofNat {a b n : ℕ} [n.AtLeastTwo] :
+    emultiplicity a b = (no_index (OfNat.ofNat n) : ℕ∞) ↔ a ^ n ∣ b ∧ ¬a ^ (n + 1) ∣ b :=
```

Comment:
Do we need `no_index` here if we are not adding this a `simp` theorem?

Reply:
Aha, interesting, seems like not, removed. Thanks!

Reply:
It could still be used with `simp [enultiplicity_eq_ofNat]`, so maybe it should be kept

Reply:
Now that I looked, I found a decent number of similar non-simp lemmas with it now (e.g. `floor_add_ofNat`, `ceil_add_ofNat`, `PosSemidef.ofNat`, ...) so I've put it back, presumably for the reason you mention, though I'm obviously happy to follow whatever you tell me to do if preferred.

Reply:
Note that you don't use `ofNat` in the RHS. What do you get (full term, not pretty-printed), if you (a) `rw`; (b) `simp` using this lemma?

Reply:
Thanks Yury.

<details>

<summary>I get (before and after the rewrite)</summary>

```
⊢ And (@Dvd.dvd.{0} Nat Nat.instDvd (@OfNat.ofNat.{0} Nat 9 (instOfNatNat 9)) (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18)))
  (Not
    (@Dvd.dvd.{0} Nat Nat.instDvd (@OfNat.ofNat.{0} Nat 27 (instOfNatNat 27))
      (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18))))

▶ expected type (10:9-10:31)
⊢ ∀ {a b n : Nat} [inst : Nat.AtLeastTwo n],
  Iff
    (@Eq.{1} ENat (@emultiplicity.{0} Nat Nat.instMonoid a b)
      (@OfNat.ofNat.{0} ENat n (@instOfNatAtLeastTwo.{0} ENat n ENat.instNatCast inst)))
    (And
      (@Dvd.dvd.{0} Nat Nat.instDvd
        (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid)) a n) b)
      (Not
        (@Dvd.dvd.{0} Nat Nat.instDvd
          (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid)) a
            (@HAdd.hAdd.{0, 0, 0} Nat Nat Nat (@instHAdd.{0} Nat instAddNat) n
              (@OfNat.ofNat.{0} Nat 1 (instOfNatNat 1))))
          b)))

-- /Users/julian/Development/mathlib4/foo.lean at 5:3
⊢ @Eq.{1} ENat
  (@emultiplicity.{0} Nat Nat.instMonoid (@OfNat.ofNat.{0} Nat 3 (instOfNatNat 3))
    (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18)))
  (@OfNat.ofNat.{0} ENat 2
    (@instOfNatAtLeastTwo.{0} ENat 2 ENat.instNatCast (@instNatAtLeastTwo (@OfNat.ofNat.{0} Nat 0 (instOfNatNat 0)))))

-- /Users/julian/Development/mathlib4/foo.lean at 5:29
⊢ And
  (@Dvd.dvd.{0} Nat Nat.instDvd
    (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid))
      (@OfNat.ofNat.{0} Nat 3 (instOfNatNat 3)) 2)
    (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18)))
  (Not
    (@Dvd.dvd.{0} Nat Nat.instDvd
      (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid))
        (@OfNat.ofNat.{0} Nat 3 (instOfNatNat 3))
        (@HAdd.hAdd.{0, 0, 0} Nat Nat Nat (@instHAdd.{0} Nat instAddNat) 2 (@OfNat.ofNat.{0} Nat 1 (instOfNatNat 1))))
      (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18))))

▶ expected type (5:7-5:29)
⊢ ∀ {a b n : Nat} [inst : Nat.AtLeastTwo n],
  Iff
    (@Eq.{1} ENat (@emultiplicity.{0} Nat Nat.instMonoid a b)
      (@OfNat.ofNat.{0} ENat n (@instOfNatAtLeastTwo.{0} ENat n ENat.instNatCast inst)))
    (And
      (@Dvd.dvd.{0} Nat Nat.instDvd
        (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid)) a n) b)
      (Not
        (@Dvd.dvd.{0} Nat Nat.instDvd
          (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid)) a
            (@HAdd.hAdd.{0, 0, 0} Nat Nat Nat (@instHAdd.{0} Nat instAddNat) n
              (@OfNat.ofNat.{0} Nat 1 (instOfNatNat 1))))
          b)))

-- /Users/julian/Development/mathlib4/foo.lean at 10:3
⊢ @Eq.{1} ENat
  (@emultiplicity.{0} Nat Nat.instMonoid (@OfNat.ofNat.{0} Nat 3 (instOfNatNat 3))
    (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18)))
  (@OfNat.ofNat.{0} ENat 2
    (@instOfNatAtLeastTwo.{0} ENat 2 ENat.instNatCast (@instNatAtLeastTwo (@OfNat.ofNat.{0} Nat 0 (instOfNatNat 0)))))

-- /Users/julian/Development/mathlib4/foo.lean at 10:31
⊢ And (@Dvd.dvd.{0} Nat Nat.instDvd (@OfNat.ofNat.{0} Nat 9 (instOfNatNat 9)) (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18)))
  (Not
    (@Dvd.dvd.{0} Nat Nat.instDvd (@OfNat.ofNat.{0} Nat 27 (instOfNatNat 27))
      (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18))))

▶ expected type (10:9-10:31)
⊢ ∀ {a b n : Nat} [inst : Nat.AtLeastTwo n],
  Iff
    (@Eq.{1} ENat (@emultiplicity.{0} Nat Nat.instMonoid a b)
      (@OfNat.ofNat.{0} ENat n (@instOfNatAtLeastTwo.{0} ENat n ENat.instNatCast inst)))
    (And
      (@Dvd.dvd.{0} Nat Nat.instDvd
        (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid)) a n) b)
      (Not
        (@Dvd.dvd.{0} Nat Nat.instDvd
          (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid)) a
            (@HAdd.hAdd.{0, 0, 0} Nat Nat Nat (@instHAdd.{0} Nat instAddNat) n
              (@OfNat.ofNat.{0} Nat 1 (instOfNatNat 1))))
          b)))
```

</details>

If I change the RHS to be:

```lean
    emultiplicity a b = (no_index (OfNat.ofNat n) : ℕ∞) ↔
      a ^ (no_index (OfNat.ofNat n) : ℕ) ∣ b ∧ ¬a ^ ((no_index (OfNat.ofNat n) : ℕ) + 1) ∣ b :=
```

<details>

<summary>I instead get</summary>

```

-- /Users/julian/Development/mathlib4/foo.lean at 5:3
⊢ @Eq.{1} ENat
  (@emultiplicity.{0} Nat Nat.instMonoid (@OfNat.ofNat.{0} Nat 3 (instOfNatNat 3))
    (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18)))
  (@OfNat.ofNat.{0} ENat 2
    (@instOfNatAtLeastTwo.{0} ENat 2 ENat.instNatCast (@instNatAtLeastTwo (@OfNat.ofNat.{0} Nat 0 (instOfNatNat 0)))))

-- /Users/julian/Development/mathlib4/foo.lean at 5:29
⊢ And
  (@Dvd.dvd.{0} Nat Nat.instDvd
    (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid))
      (@OfNat.ofNat.{0} Nat 3 (instOfNatNat 3)) (@OfNat.ofNat.{0} Nat 2 (instOfNatNat 2)))
    (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18)))
  (Not
    (@Dvd.dvd.{0} Nat Nat.instDvd
      (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid))
        (@OfNat.ofNat.{0} Nat 3 (instOfNatNat 3))
        (@HAdd.hAdd.{0, 0, 0} Nat Nat Nat (@instHAdd.{0} Nat instAddNat) (@OfNat.ofNat.{0} Nat 2 (instOfNatNat 2))
          (@OfNat.ofNat.{0} Nat 1 (instOfNatNat 1))))
      (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18))))

▶ expected type (5:7-5:29)
⊢ ∀ {a b n : Nat} [inst : Nat.AtLeastTwo n],
  Iff
    (@Eq.{1} ENat (@emultiplicity.{0} Nat Nat.instMonoid a b)
      (@OfNat.ofNat.{0} ENat n (@instOfNatAtLeastTwo.{0} ENat n ENat.instNatCast inst)))
    (And
      (@Dvd.dvd.{0} Nat Nat.instDvd
        (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid)) a
          (@OfNat.ofNat.{0} Nat n (instOfNatNat n)))
        b)
      (Not
        (@Dvd.dvd.{0} Nat Nat.instDvd
          (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid)) a
            (@HAdd.hAdd.{0, 0, 0} Nat Nat Nat (@instHAdd.{0} Nat instAddNat) (@OfNat.ofNat.{0} Nat n (instOfNatNat n))
              (@OfNat.ofNat.{0} Nat 1 (instOfNatNat 1))))
          b)))

-- /Users/julian/Development/mathlib4/foo.lean at 10:3
⊢ @Eq.{1} ENat
  (@emultiplicity.{0} Nat Nat.instMonoid (@OfNat.ofNat.{0} Nat 3 (instOfNatNat 3))
    (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18)))
  (@OfNat.ofNat.{0} ENat 2
    (@instOfNatAtLeastTwo.{0} ENat 2 ENat.instNatCast (@instNatAtLeastTwo (@OfNat.ofNat.{0} Nat 0 (instOfNatNat 0)))))

-- /Users/julian/Development/mathlib4/foo.lean at 10:31
⊢ And (@Dvd.dvd.{0} Nat Nat.instDvd (@OfNat.ofNat.{0} Nat 9 (instOfNatNat 9)) (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18)))
  (Not
    (@Dvd.dvd.{0} Nat Nat.instDvd (@OfNat.ofNat.{0} Nat 27 (instOfNatNat 27))
      (@OfNat.ofNat.{0} Nat 18 (instOfNatNat 18))))

▶ expected type (10:9-10:31)
⊢ ∀ {a b n : Nat} [inst : Nat.AtLeastTwo n],
  Iff
    (@Eq.{1} ENat (@emultiplicity.{0} Nat Nat.instMonoid a b)
      (@OfNat.ofNat.{0} ENat n (@instOfNatAtLeastTwo.{0} ENat n ENat.instNatCast inst)))
    (And
      (@Dvd.dvd.{0} Nat Nat.instDvd
        (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid)) a
          (@OfNat.ofNat.{0} Nat n (instOfNatNat n)))
        b)
      (Not
        (@Dvd.dvd.{0} Nat Nat.instDvd
          (@HPow.hPow.{0, 0, 0} Nat Nat Nat (@instHPow.{0, 0} Nat Nat (@Monoid.toNatPow.{0} Nat Nat.instMonoid)) a
            (@HAdd.hAdd.{0, 0, 0} Nat Nat Nat (@instHAdd.{0} Nat instAddNat) (@OfNat.ofNat.{0} Nat n (instOfNatNat n))
              (@OfNat.ofNat.{0} Nat 1 (instOfNatNat 1))))
          b)))


```

</details>

I'm not 100% sure what I'm looking for (and I see a few examples of lemmas that also do not have `OfNat` on the RHS, e.g. [this one](https://github.com/leanprover-community/mathlib4/blob/b0ce30989a48fd07c7e13c64152865e11a65c6b4/Mathlib/Algebra/Polynomial/Eval/Defs.lean#L401). I see a small diff between the two terms obviously. Can you help me understand what I should be seeing there in terms of red flags (or lack thereof)?

Reply:
Can you give a mwe that shows exactly how you are using it?

Reply:
> and I see a few examples of lemmas that also do not have `OfNat` on the RHS, e.g. [this one](https://github.com/leanprover-community/mathlib4/blob/b0ce30989a48fd07c7e13c64152865e11a65c6b4/Mathlib/Algebra/Polynomial/Eval/Defs.lean#L401).

This lemma is mis-stated, `ofNat` should indeed be on the RHS.

`no_index` matters a little less, and I think we often skip it on the RHS out of laziness. Ideally Lean would change to make it unnecessary.

Reply:
> Can you give a mwe that shows exactly how you are using it?


The example I was using to confirm this "works" (where my understanding of works is probably limited to "does not require `erw`) was the one from the Zulip thread, i.e.:

```lean
import Mathlib.RingTheory.Multiplicity

example : emultiplicity 3 18 = 2 := by
  rw [emultiplicity_eq_ofNat]
  decide
```

> This lemma is mis-stated, ofNat should indeed be on the RHS.

(I believe you of course, I'll just note there are not just one of them, there were at least a handful of lemmas like that to my recollection, so sounds like there might be a few places in need of cleanup. I don't feel comfortable doing so until I understand what any of this means, maybe after your response!)